### PR TITLE
[New APIs] Network IP Availability and Usage Stats

### PR DESCRIPTION
**Pull request by kevin-zhonghao**
This PR proposes the following changes:
* Design and implement new APIs for Network IP Availability and Usage Stats
* Refactor Subnet/VPC/Segments/SegmentRanges

### CODE REVIEW

File: web/src/main/java/com/futurewei/alcor/web/entity/vpc/NetworkIPAvailabilityEntity.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+*/
+package com.futurewei.alcor.web.entity.vpc;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+public class NetworkIPAvailabilityEntity {
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/entity/vpc/NetworkIPAvailabilityEntity.java:24]: The class name is internal so I would suggest us to use VpcIpAvailabilityEntity. Not a big deal. Your call.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/utils/IPUtil.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                                    count ++;
+                                }
+                            } else {
+                                for (ip4 = 0; ip4 < 256; ip4++) {
+                                    count ++;
+                                }
+                            }
+                        }
+                    } else {
+                        for (ip3 = 0; ip3 < 256; ip3++) {
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/utils/IPUtil.java:46]: May I know why it is 256?

Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/utils/IPUtil.java:46]: The range of each segment of the IP address is 0-255. The else{...} here mean we should calculate the whole range from 0 to 255

File: services/port_manager/src/main/resources/application.properties [hunk 3]
```diff
@@ -16,6 +16,6 @@ microservices.subnet.service.url=http://localhost:9002
 microservices.route.service.url=http://localhost:9003/routes
 microservices.ip.service.url=http://localhost:9004/ips
 microservices.mac.service.url=http://localhost:9005/macs
-microservices.sg.service.url=http://127.0.0.1:9006/securitygroups
+microservices.sg.service.url=http://localhost:9008/securitygroups
 microservices.node.service.url=http://localhost:9007/nodes
-microservices.dataplane.service.url=http://localhost:9100
```
Comment (xieus) [inline services/port_manager/src/main/resources/application.properties:21]: Please remove local changes in this PR as well.

**Comment by xieus** [comment]
@kevin-zhonghao Let us get back to this PR. It seems that we have quite a few merge conflicts now due to many recent changes. 

Let us go over this PR together and clean up some debt :-)

File: services/port_manager/src/main/resources/application.properties [hunk 4]
```diff
@@ -16,6 +16,6 @@ microservices.subnet.service.url=http://localhost:9002
...
\ No newline at end of file
+microservices.dataplane.service.url=http://10.213.43.187:8080
```
Comment (xieus) [inline services/port_manager/src/main/resources/application.properties:21]: We have a recent change on the DPM port. We will likely need to merge from latest master.

File: sh.exe.stackdump [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Stack trace:
```
Comment (xieus) [inline sh.exe.stackdump:1]: Please remove local file.

btw, this means that we need to add one more item to our .gitignore file.

Comment (kevin-zhonghao) [reply sh.exe.stackdump:1]: Sure

**Pull request closed without merging**